usb/ehci/phy-ia32-generic.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Phoenix-RTOS
+ *
+ * EHCI USB Physical Layer for ia32
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Adam Greloch
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/platform.h>
+#include <phoenix/arch/ia32/ia32.h>
+#include <stdio.h>
+#include <sys/platform.h>
+
+#include <hcd.h>
+
+
+#define EHCI_MAP_SIZE (0x1000)
+#define EHCI_PROGIF   (0x20)
+
+
+int hcd_getInfo(const hcd_info_t **info)
+{
+	platformctl_t pctl;
+	int i, err;
+
+	hcd_info_t *hcd_info = NULL;
+
+	pctl.action = pctl_get;
+	pctl.type = pctl_pci;
+	pctl.pci.id.cl = 0x0c03;
+	pctl.pci.id.progif = EHCI_PROGIF;
+	pctl.pci.dev.bus = 0;
+	pctl.pci.dev.dev = 0;
+	pctl.pci.dev.func = 0;
+	pctl.pci.caps = NULL;
+
+	i = 0;
+	for (;;) {
+		pctl.pci.id.vendor = PCI_VENDOR_INTEL;
+		pctl.pci.id.device = PCI_ANY;
+		pctl.pci.id.subvendor = PCI_ANY;
+		pctl.pci.id.subdevice = PCI_ANY;
+
+		err = platformctl(&pctl);
+
+		if (err == -ENODEV) {
+			break;
+		}
+
+		if (err < 0) {
+			fprintf(stderr, "phy: pctl_get failed\n");
+			return err;
+		}
+
+		if (pctl.pci.dev.progif != EHCI_PROGIF) {
+			pctl.pci.dev.func++;
+			continue;
+		}
+
+		hcd_info = realloc(hcd_info, sizeof(hcd_info_t) * (i + 1));
+		if (hcd_info == NULL) {
+			return -ENOMEM;
+		}
+		memset((void *)(hcd_info + i), 0, sizeof(hcd_info_t));
+
+		sprintf(hcd_info[i].type, "ehci");
+		hcd_info[i].hcdaddr = pctl.pci.dev.resources[0].base;
+
+		/* TODO do ACPI _PRT lookup instead */
+		fprintf(stderr, "phy: choosing default irq from pci\n");
+		hcd_info[i].irq = pctl.pci.dev.irq;
+
+		hcd_info[i].pci_devId.bus = pctl.pci.dev.bus;
+		hcd_info[i].pci_devId.dev = pctl.pci.dev.dev;
+		hcd_info[i].pci_devId.func = pctl.pci.dev.func;
+
+		i++;
+		pctl.pci.dev.func++;
+	}
+
+	*info = hcd_info;
+
+	return i;
+}
+
+
+static void phy_config(hcd_t *hcd)
+{
+	platformctl_t pctl;
+	int err;
+	uint8_t eecp;
+	short bus = hcd->info->pci_devId.bus;
+	short dev = hcd->info->pci_devId.dev;
+	short func = hcd->info->pci_devId.func;
+
+	pctl.action = pctl_set;
+	pctl.type = pctl_pcicfg;
+	pctl.pcicfg.dev.bus = bus;
+	pctl.pcicfg.dev.dev = dev;
+	pctl.pcicfg.dev.func = func;
+	pctl.pcicfg.cfg = pci_cfg_busmaster;
+	pctl.pcicfg.enable = 1;
+
+	err = platformctl(&pctl);
+	if (err < 0) {
+		fprintf(stderr, "phy: setting busmaster failed: %d\n", err);
+	}
+
+	pctl.pcicfg.cfg = pci_cfg_memoryspace;
+	pctl.pcicfg.enable = 1;
+
+	err = platformctl(&pctl);
+	if (err < 0) {
+		fprintf(stderr, "phy: setting memoryspace failed: %d\n", err);
+	}
+
+	pctl.pcicfg.cfg = pci_cfg_interruptdisable;
+	pctl.pcicfg.enable = 0;
+
+	err = platformctl(&pctl);
+	if (err < 0) {
+		fprintf(stderr, "phy: enabling interrupts failed: %d\n", err);
+	}
+
+	eecp = (*(uint32_t *)((char *)hcd->base + 0x8) >> 8) & 0xFF;
```

Comment:
Is `+ 0x8` a specific register's offset? If so, maybe add an enum for it?

Reply:
It's an offset inside the HCCPARAMS register pointing to EECP. Added a local define, as this pointer is unlikely to be used anywhere else than this file